Add tests to CPX variants in SV pipeline and fixing bugs

SHuang-Broad opened the pull request:
E.g. The `CTG_NAMES` and `TOTAL_MAPPINGS`.

Also grouping the equivalent variant from different contigs into one, i.e. avoiding duplicates.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/CpxVariantDetector.java:
```diff
@@ -44,7 +44,8 @@
  * but that could be dealt with later.
  */
-public final class CpxVariantDetector {
+public final class CpxVariantDetector implements Serializable {
```
Why is this `java.io.Serializable`? I don't see where this class get serialized.. I always get nervous when we mix java serialization with kryo serialization.

cwhelan replied:
If it's for some reason necessary because you're serializing the inner class, might be better to bump the inner class to a top-level class.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/CpxVariantDetector.java:
```diff
@@ -879,6 +852,153 @@ private static boolean alignmentIsDisjointFromAlphaOmega(final SimpleInterval al
+     * @return
+     */
+    @DefaultSerializer(EquivKey.Serializer.class)
+    private static final class EquivKey {
```
Can you give this class a name that is more descriptive of what it represents? I'm thinking something like `CpxVariantCanonicalRepresentation` or something like that. Make sense to you?

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/CpxVariantDetector.java:
```diff
@@ -59,8 +60,17 @@ public void inferSvAndWriteVCF(final JavaRDD<AssemblyContigWithFineTunedAlignmen
+                annotatedContigs
+                        .groupBy(tig -> new EquivKey(tig.getAffectedRefRegion(), tig.getAltSeq(), tig.getSegmentsAndDescription()))
+                        .map(pair -> {
+                            final VariantContext rawVariantContext = pair._1.toVariantContext(referenceBroadcast.getValue());
```
I think it might be nice to extract this into a method called `produceCpxVariantContext(Tuple2<>...)`. Then it would be great to write a little unit test demonstrating the functionality (ie getting one variant context back with the right annotations set).

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/CpxVariantDetector.java:
```diff
@@ -879,6 +852,153 @@ private static boolean alignmentIsDisjointFromAlphaOmega(final SimpleInterval al
+        attributeMap.put(GATKSVVCFConstants.TOTAL_MAPPINGS,    String.valueOf(sorted.size()));
+        attributeMap.put(GATKSVVCFConstants.CONTIG_NAMES,      sorted.stream().map(annotatedContig -> annotatedContig.tigWithInsMappings.getSourceContig().contigName).collect(Collectors.joining(VCFConstants.INFO_FIELD_ARRAY_SEPARATOR)));
+
+        // TODO: 2/1/18 question: show we introduce new annotations for them?
```
What do you mean by "introduce new annotations"? I'd add these in so that your calls are equivalent to the calls from the current tool. However, to make the annotations for these complex calls equivalent to those for deletions, etc, I'd look only at the flanking alignments that anchor the variant to the reference -- ie. the head and tail alignments from the contig. Does that make sense?

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/CpxVariantDetector.java:
```diff
@@ -879,6 +852,153 @@ private static boolean alignmentIsDisjointFromAlphaOmega(final SimpleInterval al
+//        attributeMap.put(GATKSVVCFConstants.ALIGN_LENGTHS,     tigWithInsMappings.getSourceContig().alignmentIntervals.stream().map(ai -> String.valueOf(ai.getSizeOnRead())).collect(Collectors.joining(VCFConstants.INFO_FIELD_ARRAY_SEPARATOR)));
+//        attributeMap.put(GATKSVVCFConstants.MAX_ALIGN_LENGTH,  String.valueOf(tigWithInsMappings.getSourceContig().alignmentIntervals.stream().mapToInt(AlignmentInterval::getSizeOnRead).max().orElse(0)));
+
+        // TODO: 12/11/17 integrate these with those that survived the alignment filtering step?
```
I'm a little confused about what would go into this annotation for a complex variant. Don't you already cover this with your segments representation? Maybe you could write a unit test with some examples?

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/CpxVariantDetector.java:
```diff
@@ -879,6 +852,153 @@ private static boolean alignmentIsDisjointFromAlphaOmega(final SimpleInterval al
+            return attributes;
+        }
+
+        VariantContext toVariantContext(final ReferenceMultiSource reference) throws IOException {
```
What if you made this return a `VariantContextBuilder` instead of a `VariantContext`, then you wouldn't have to make it, then make a new builder from the new `VariantContext`, and then make that?

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/CpxVariantDetector.java:
```diff
@@ -879,6 +852,153 @@ private static boolean alignmentIsDisjointFromAlphaOmega(final SimpleInterval al
+        VariantContext toVariantContext(final ReferenceMultiSource reference) throws IOException {
+
+            final CpxVariantType cpxVariant = new CpxVariantType(affectedRefRegion, typeSpecificExtraAttributes());
+            final SimpleInterval pos = new SimpleInterval(affectedRefRegion.getContig(), affectedRefRegion.getStart(), affectedRefRegion.getStart());
```
Are we sure we want the end to be equal to the start for these?

SHuang-Broad commented:
Hi @cwhelan , I've expanded this PR to do more than what it originally was trying to fix, and separated the patches by commits as usual:

* the originally proposed fix, which brings back the annotation that are available to simple variants but go missing due to a careless bug, is now done in commit 50f1b640a31ddb528dc763b83b26a9d98dce8556; this commit also accordingly refactors the giant class `CpxVariantDetector` into three new classes
* in the 2nd commit 734516383fb665a79796de76535560fc03cb754b, I did more refactoring on how we group the descriptions for the annotation keys, and updated the test VCF files accordingly.
* because of the refactoring, the review comments were gone, so I added them back in the 3rd commit b7619c45a949dfba21d65a5ed876bc72e832aa77, which contains the comments and my replies. They come in as TODO's but are going to be removed ultimately
* in the following commits, I added tests for the CPX code path, selecting three representative cases (there's no limit how complex the scenario can go). One particular commit 224c97c7b736e94ed6b4d8b067ec830a9f8f2403 is large but most of it is for adding a flat file that contains the chromosome names in hg38 and their lengths for building a bare bone sequence dictionary used in building test data.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/CpxVariantCanonicalRepresentation.java:
```diff
@@ -348,10 +362,33 @@ static SimpleInterval getAffectedReferenceRegion(final List<SimpleInterval> even
+         * the variable "pos" below is going to be the POS column of the VCF record,
+         * and several lines down below is where we populate the END INFO field.
+         */
         final SimpleInterval pos = new SimpleInterval(affectedRefRegion.getContig(), affectedRefRegion.getStart(), affectedRefRegion.getStart());
```
Oh I see. I was confused about what this interval is being used for, sorry. If the variable was named 'refAlleleInterval' or something maybe it would help?

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/CpxVariantInterpreter.java:
```diff
@@ -113,18 +121,153 @@ private static AssemblyContigWithFineTunedAlignments furtherPreprocess(final Ass
+            } else {
+                Boolean hintRequestYieldingOverlapToAlignmentTwo = null;
+                if (one.equals(originalAlignments.get(0))) {
+                    hintRequestYieldingOverlapToAlignmentTwo = Boolean.FALSE;
```
just use `false` or `true`

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/CpxVariantInterpreter.java:
```diff
@@ -113,18 +121,153 @@ private static AssemblyContigWithFineTunedAlignments furtherPreprocess(final Ass
+                result.add(one);
+                one = two;
+            } else {
+                Boolean hintRequestYieldingOverlapToAlignmentTwo = null;
```
Rather than a three value variable (`null`/`true`/`false`) which is effectively how you're using this, I think it would be more explicit if you just had two booleans, eg, `firstIsAlignmentHead`, `secondIsAlignmentTail`

SHuang-Broad replied:
changed as suggested.

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/CpxVariantInterpreter.java:
```diff
@@ -113,18 +121,153 @@ private static AssemblyContigWithFineTunedAlignments furtherPreprocess(final Ass
+
+        final AlignmentInterval head = result.get(0);
+        final AlignmentInterval tail = result.get(result.size() - 1);
+        final int ovp = AlignmentInterval.overlapOnContig(head, tail);
```
spell out `overlap`

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/CpxVariantCanonicalRepresentation.java:
```diff
@@ -100,6 +98,31 @@
+    @VisibleForTesting
+    static CpxVariantCanonicalRepresentation makeInterpretation(final CpxVariantInducingAssemblyContig cpxVariantInducingAssemblyContig) {
+
+        if (cpxVariantInducingAssemblyContig.getEventPrimaryChromosomeSegmentingLocations().size() == 1) {
```
Why have two different code paths here and two different constructors? All this method is doing is looking at data in the `cpxVariantInducingAssemblyContig`, so why not just have one constructor that takes that? Then this method can be super simple or just go away.

SHuang-Broad replied:
changed as suggested. now much cleaner. Thanks!

cwhelan commented on src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/inference/CpxVariantCanonicalRepresentation.java:
```diff
@@ -447,8 +467,12 @@ private CpxVariantCanonicalRepresentation(final Kryo kryo, final Input input) {
         referenceSegments = new ArrayList<>(numSegments);
-        for (int i = 0; i < numSegments; ++i)
-            referenceSegments.add(kryo.readObject(input, SimpleInterval.class));
+        for (int i = 0; i < numSegments; ++i) {
```
Why did you have to re-do this serialization instead of delegating to `SimpleInterval.Serializer`?

SHuang-Broad replied:
There was no Kryo serializer for `SimpleInterval` since it is Java serializable, and the class is not registered either.

The serialization tests (which are new in this PR) fail if I did the `kryo.readObject(input, SimpleInterval.class)`
